packages/data-broker/src/services/dhis/DhisInputSchemeResolvingApi.js

```diff
@@ -0,0 +1,306 @@
+/**
+ * Tupaia
+ * Copyright (c) 2017 - 2020 Beyond Essential Systems Pty Ltd
+ */
+
+// Note: should really extend DhisApi but there's some sort of circular dependency issue
+
+// TODO: move to actual db table
+const TABLE_DHIS_ORG_UNIT_MAPPING = {
+  TO: 'jSk4k1gnSqk',
+};
+
+// TODO: move to actual db table
+const DATA_SOURCE_CONFIG_BY_CODE = {
+  // SS162: { dhisId: 'pDhCn9lbi5v' },
+  // SS128: { dhisId: 'K3wJsRJHMCa' },
+  // SS182: { dhisId: 'g2KqfrYXLdM' },
+  // SS190A: { dhisId: 'iYzYjNzDnAY' },
+  // SS192: { dhisId: 'RngB7fLGI3i' },
+  // SS219: { dhisId: 'bcg0ApKZTav' },
+  // DP9: { dhisId: 'OzDb71SXJgP' },
+  HP02: { dhisId: 'OmssK27WEJB' },
+};
+
+export class DhisInputSchemeResolvingApi {
+  constructor(models, api) {
+    this.models = models;
+    this.api = api;
+  }
+
+  async getAnalytics(query) {
+    let modifiedQuery = { ...query };
+
+    // console.log('DhisInputSchemeResolvingApi', 'getAnalytics', 'query', query);
+
+    if (
+      (await this.anyDataElementsHaveDhisId(query)) ||
+      (await this.anyOrgUnitsHaveDhisId(query))
+    ) {
+      // the endpoint used /api/analytics/rawData.json only allows a single "inputIdScheme", which means
+      // both the dataElements and orgUnits need to be ids
+
+      modifiedQuery = await this.replaceDataElementCodesWithIds(modifiedQuery);
+      modifiedQuery = await this.replaceOrgUnitCodesWithIds(modifiedQuery);
+
+      modifiedQuery.inputIdScheme = 'uid';
+    }
+
+    // console.log('DhisInputSchemeResolvingApi', 'getAnalytics', 'modifiedQuery', modifiedQuery);
+    return this.api.getAnalytics(modifiedQuery);
+  }
+
+  async getEventAnalytics(query) {
+    // console.log('DhisInputSchemeResolvingApi', 'getEventAnalytics', 'query', query);
```

Comment:
The base method here, DhisApi#getEventAnalytics will convert codes to ids before submitting the actual data request. It does that with separate calls e.g. /api/dataElements. To prevent this (because laos does not have any codes), we pre-emptively swap out codes for ids, using our internal mapping.

Reply:
Added as comment